openedx/adg/admin_dashboard/performance/views.py

```diff
@@ -1,3 +1,151 @@
+
+        user_ids = [user.id for user in context_data['object_list']]
+        course_grades = get_course_grades_for_users(user_ids, self.course.id)
+        for user in context_data['object_list']:
+            user.course_grade = course_grades[user.id]
+
+        return context_data
+
+    def get(self, request, *args, **kwargs):
+        """
+        Overridden to add extra context data.
+        """
+        self.extra_context = {
+            'course_name': self.course.display_name,
+        }
+        return super().get(request, *args, **kwargs)
+
+    def post(self, request, *args, **kwargs):
+        """
+        Attaches csv report for download with response.
+
+        Returns:
+            HttpResponse: Response with grades report.
+        """
+        users = self.get_queryset().only('id', 'username', 'email', 'profile__phone_number', 'profile__name',)
+        course_grades = get_course_grades_for_users(users.values_list('id'), self.course.id)
+
+        report_data = get_course_grades_report_data(users, course_grades)
+        buf = StringIO()
+        writer = csv.writer(buf)
+        writer.writerows(report_data)
+
+        response = HttpResponse(buf.getvalue(), content_type='text/csv')
+        response['Content-Disposition'] = f'attachment; filename={self.course.display_name}_{date.today()}_grades.csv'
+        return response
+
+    def get_queryset(self):
+        """
+        Overridden to custom filters.
+
+        Returns
+            queryset: Queryset containing enrolled users of the course.
+        """
+        course_enrollments_prefetch_object = Prefetch(
+            'courseenrollment_set',
+            queryset=CourseEnrollment.objects.filter(
+                course_id=self.course.id).only(
+                'id', 'created', 'user_id'
+            ),
+            to_attr='course_enrollment'
+        )
+        enrolled_students = User.objects.filter(
+            courseenrollment__course_id=self.course.id,
+            courseenrollment__is_active=1
+        ).select_related('profile').prefetch_related(course_enrollments_prefetch_object)
+
+        query = self.request.GET.get('q', None)
+        if query:
+            enrolled_students = enrolled_students.filter(
+                Q(profile__name__icontains=query) | Q(email=query)
+            )
+
+        return enrolled_students
+
+    def get_paginate_by(self, queryset):
+        """
+        Overridden to show custom number of results per page.
+
+        Returns:
+            int: Integer for pagination results per page.
+        """
+        return self.request.GET.get('paginate_by', self.DEFAULT_COURSE_GRADES_PAGINATE_BY)
+
+    def _get_course(self):
+        """
+        Fetches the course using the URL kwarg.
+
+        Returns:
+            Course: Fetched course.
+        """
+        course_key_string = self.kwargs['course_key_string']
+        course_key = CourseKey.from_string(course_key_string)
+        return get_course_by_id(course_key)
```

Comment:
Actually it is a list view and the above mentioned mixin is used for the detail views.

Reply:
We can still add that mixin in our class and see if we can achieve our functionality. It is added to DetailView by default.

Reply:
SingleObjectMixin is useful if we can define a queryset from which it can get the object using the URL kwarg but in our case we have Users queryset. Also if I am to use this mixin I will have to override the get_object and will have to define custom queryset as the current queryset returns Users. This will destroy whole purpose of using the SingleObjectMixin, Also this is used to get the details for an object but we are not showing details of course.